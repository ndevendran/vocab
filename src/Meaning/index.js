import React from 'react';
import {Button} from 'react-bootstrap';

function Meaning({definition, hidden, onClick}) {
  return (
    <div>
      {hidden
       ? <Button className="showDefinition" variant="info" onClick={onClick}>Show Definition</Button>
       : <span>{definition}</span>
      }
    </div>
  );
}

export default Meaning;
