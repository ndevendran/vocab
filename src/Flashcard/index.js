import React, {Component} from 'react';
import {Button} from 'react-bootstrap';
import Meaning from '../Meaning';
import PropTypes from 'prop-types';

Button.propTypes = {
  onClick: PropTypes.func
};

Meaning.propTypes = {
  onClick: PropTypes.func,
  definition: PropTypes.string,
  hidden: PropTypes.bool
};


class Flashcard extends Component {
  constructor(props) {
    super(props);

    this.state = {
      index: 0,
      hideMeaning: true,
    };

    this.getPreviousWord = this.getPreviousWord.bind(this);
    this.getNextWord = this.getNextWord.bind(this);
    this.showDefinition = this.showDefinition.bind(this);
  }

  showDefinition() {
    const hideMeaning = !this.state.hideMeaning;
    this.setState({hideMeaning: hideMeaning});
  }

  getNextWord() {
      this.setState((prevState) => {
        let newIndex;
        if((prevState.index + 1) === this.props.words.length) {
          newIndex = 0;
        } else {
          newIndex = prevState.index+1;
        }
        return {index: newIndex, hideMeaning: true};
      });
  }

  getPreviousWord() {
    this.setState((prevState) => {
      let newIndex;
      if(prevState.index === 0) {
        newIndex = this.props.words.length - 1;
      } else {
        newIndex = prevState.index - 1;
      }
      return {index: newIndex, hideMeaning: true};
    });
  }

  render () {
    const hideMeaning = this.state.hideMeaning;
    const word = this.props.words[this.state.index];
    return (
      <div className="flashcard">

          <div>
            <div className="word">Word: {word.name}</div>
            <Meaning definition={word.definition}
              hidden={hideMeaning} onClick={this.showDefinition}
            />
            <div className="interactions">
              <Button variant="primary" onClick={this.getPreviousWord}>Prev</Button>
              <Button variant="primary" onClick={this.getNextWord}>Next</Button>
            </div>
        </div>
      </div>
    );
  }
}

export default Flashcard;
