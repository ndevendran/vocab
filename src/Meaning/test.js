import React from 'react';
import ReactDOM from 'react-dom';
import Meaning from './Meaning';

it('renders without crashing', () => {
  const div = document.createElement('div');
  const props = {
    definition: "test",
    hidden: false,
    onClick: () => true,
  }
  ReactDOM.render(<Meaning ...props />, div);
  ReactDOM.unmountComponentAtNode(div);
});
