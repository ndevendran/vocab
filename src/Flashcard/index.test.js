import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Flashcard from '../Flashcard';
import Meaning from '../Meaning';

Enzyme.configure({ adapter: new Adapter() });

describe('Flashcard', () => {
  const props = {
    words: [{"name": "test", "definition": "test"}],
    index: 0,
    hideMeaning: true,
    hideStep: true
  }

  it('loads without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(<Flashcard {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('snapshot is valid', () => {
    const component = renderer.create(
      <Flashcard {...props} />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders one word when step is not hidden', () => {
    const customProps = {
      words: [{"name": "test", "definition": "custom test"}],
      hideStep: false,
      hideMeaning: false,
      index: 0
    };
    const wrapper = shallow(<Flashcard {...customProps} />);
    expect(wrapper.find('.word').length).toBe(1);
  });

  it('renders no words when step is hidden', () => {
    const customProps = {
      words: [{"name": "test", "definition": "custom test"}],
      hideStep: true,
      hideMeaning: false,
      index: 0
    };
    const wrapper = shallow(<Flashcard {...customProps} />);
    expect(wrapper.find('.word').length).toBe(0);
  });

  it('Contains a Meaning component', () => {
    const customProps = {
      words: [{"name": "test", "definition": "custom test"}],
      hideStep: false,
      hideMeaning: false,
      index: 0
    };
    const wrapper = shallow(<Flashcard {...customProps} />);

    const word = customProps.words[customProps.index];
    expect(wrapper.find(Meaning)).toHaveLength(1);
  });
});
