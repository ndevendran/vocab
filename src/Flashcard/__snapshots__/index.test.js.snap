// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flashcard snapshot is valid 1`] = `
<div
  className="flashcard"
>
  <div>
    <div
      className="word"
    >
      Word: 
      test
    </div>
    <div>
      <button
        className="showDefinition btn btn-info"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Show Definition
      </button>
    </div>
    <div
      className="interactions"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Prev
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
